ssh/keygen.go

```diff
@@ -0,0 +1,119 @@
+package ssh
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os/exec"
+	"path"
+	"strconv"
+
+	"github.com/spf13/pflag"
+)
+
+// OptionalValue is an extension of pflag.Value that remembers if it has been
+// set.
+type OptionalValue interface {
+	pflag.Value
+	Specified() bool
+}
+
+// KeyBitsValue is an OptionalValue allowing specification of the -b argument
+// to ssh-keygen.
+type KeyBitsValue struct {
+	specified bool
+	keyBits   uint64
+}
+
+func (kbv *KeyBitsValue) String() string {
+	return strconv.FormatUint(kbv.keyBits, 10)
+}
+
+func (kbv *KeyBitsValue) Set(s string) error {
+	v, err := strconv.ParseUint(s, 0, 64)
+	if err != nil {
+		return err
+	}
+
+	kbv.keyBits = v
+	kbv.specified = true
+
+	return nil
+}
+
+func (kbv *KeyBitsValue) Type() string {
+	return "uint64"
+}
+
+func (kbv *KeyBitsValue) Specified() bool {
+	return kbv.specified
+}
+
+// KeyTypeValue is an OptionalValue allowing specification of the -t argument
+// to ssh-keygen.
+type KeyTypeValue struct {
+	specified bool
+	keyType   string
+}
+
+func (ktv *KeyTypeValue) String() string {
+	return ktv.keyType
+}
+
+func (ktv *KeyTypeValue) Set(s string) error {
+	if len(s) > 0 {
+		ktv.keyType = s
+		ktv.specified = true
+	}
+	return nil
+}
+
+func (ktv *KeyTypeValue) Type() string {
+	return "string"
+}
+
+func (ktv *KeyTypeValue) Specified() bool {
+	return ktv.specified
+}
+
```

Comment:
I didn't have this originally, but the code elsewhere was much uglier as a result...